gateway/config.go

```diff
@@ -2,27 +2,33 @@ package gateway
 
 import (
 	"os"
+	"time"
 
 	"gopkg.in/yaml.v2"
 )
 
 type Config struct {
-	Server      ServerConfig `yaml:"server"`
-	Admin       ServerConfig `yaml:"admin"`
-	Tenants     []Tenant     `yaml:"tenants"`
-	Distributor struct {
-		URL   string   `yaml:"url"`
-		Paths []string `yaml:"paths"`
-	} `yaml:"distributor"`
-	QueryFrontend struct {
-		URL   string   `yaml:"url"`
-		Paths []string `yaml:"paths"`
-	} `yaml:"frontend"`
+	Server        ServerConfig `yaml:"server"`
+	Admin         ServerConfig `yaml:"admin"`
+	Tenants       []Tenant     `yaml:"tenants"`
+	Distributor   Upstream     `yaml:"distributor"`
+	QueryFrontend Upstream     `yaml:"frontend"`
+}
+
+type Upstream struct {
+	URL                             string        `yaml:"url"`
+	Paths                           []string      `yaml:"paths"`
+	HTTPClientDialerTimeout         time.Duration `yaml:"http_client_dialer_timeout"`
+	HTTPClientTLSHandshakeTimeout   time.Duration `yaml:"http_client_tls_handshake_timeout"`
+	HTTPClientResponseHeaderTimeout time.Duration `yaml:"http_client_response_header_timeout"`
```

Comment:
As mentioned in [this comment](https://github.com/cortexproject/auth-gateway/pull/18#pullrequestreview-1399450604), we should also implement `http_client_timeout`. However, I was not sure about it. In this [blog](https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/) under **Client Timeouts** section, the image shows that the supported values ie. `http_client_dialer_timeout, http_client_tls_handshake_timeout, http_client_response_header_timeout`, cover almost same as `http.Client.Timeout`. I feel like the combination of these three is the same as `http.Client.Timeout`. Therefore, I hesitated to implement it.

Can I get your opinion on this @friedrichg?

Reply:
it's not exactly the same, we need to give folks the possibility to configure them all as needed. please include them